packages/core/src/components/forms/_file-upload.scss

```diff
@@ -69,21 +70,21 @@ $file-upload-button-width-large: $pt-grid-size * 8.5;
   top: 0;
   right: 0;
   left: 0;
-  margin-right: $file-upload-button-width;
   border-top-right-radius: 0;
   border-bottom-right-radius: 0;
   height: $pt-input-height;
+  padding-right: 0;
   line-height: $pt-input-height;
   color: $pt-text-color-muted;
   user-select: none;
 
-  &::after {
+  // we use ::before rather than ::after although the "Browse" button is visually
+  // positioned to the right since this allows for .pt-file-upload-input to be
+  // paired with .pt-text-overflow-ellipsis if desired.
+  &::before {
     @include pt-button();
-    position: absolute;
-    top: 0;
-    right: 0;
-    left: 100%;
-    margin-left: -$button-border-width;
+    float: right;
+    margin-left: $input-padding-horizontal - $button-border-width;
```

Comment:
hmm this is much better solved entirely with `padding` so the `::after` button is still _inside_ for `overflow` purposes and simply covers X pixels of the padding, and the text still truncates 10 before the button. 

![image](https://user-images.githubusercontent.com/464822/29197122-f866b03c-7ded-11e7-8f15-b6076e9ac4d8.png)